File: Fortran/3.5/jacobi_openacc.f90
subroutine jacobi_parallel(u, unew, n, niter)
  implicit none
  integer, intent(in) :: n, niter
  real(kind=8), dimension(n, n, n), intent(inout) :: u, unew
  integer :: i, j, k, iter

  do iter = 1, niter
    !$ACC PARALLEL LOOP COLLAPSE(3) PRIVATE(i, j, k)
    do k = 2, n - 1
      do j = 2, n - 1
        do i = 2, n - 1
          unew(i, j, k) = 0.125 * (u(i-1, j, k) + u(i+1, j, k) + u(i,
j-1, k) &
            + u(i, j+1, k) + u(i, j, k-1) + u(i, j, k+1) + u(i, j, k))
        end do
      end do
    end do
    !$ACC END PARALLEL LOOP
    u(:, :, :) = unew(:, :, :)
  end do
end subroutine jacobi_parallel

File: C/3.5/Par/jacobi_openacc.c
void jacobi3D(double ***in, double ***out, int N, int T)
{
    double ***temp;

    for (int t = 0; t < T; t++)
    {
        #pragma acc parallel loop collapse(3)
        for (int i = 1; i < N - 1; i++)
        {
            for (int j = 1; j < N - 1; j++)
            {
                for (int k = 1; k < N - 1; k++)
                {
                    out[i][j][k] = (in[i - 1][j][k] + in[i + 1][j][k] + in[i][j - 1][k] + in[i][j + 1][k] + in[i][j][k - 1] + in[i][j][k + 1]) / 6.0;
                }
            }
        }

        // Swap in and out pointers
        temp = out;
        out = in;
        in = temp;
     }
}
